extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb

```diff
@@ -0,0 +1,329 @@
+    "Vorbis" : 6 \
+}
+constant kRecordFormats is { \
+    "ThreeGPP" : 1, \
+    "MPEG-4" : 2, \
+    "AMR WB" : 4, \
+    "AAC ADTS" : 6, \
+    "WebM" : 9 \
+}
+
+private variable mRecordInput as optional String
+private variable mRecordFormat as optional String
+private variable mRecordCompressionType as optional String
+
+private handler IsAndroid() returns Boolean
+    return the operating system is "android"
+end handler
+
+handler GetOptionalProperty(in pVar as optional String, in pDefault as String) returns String
+    if pVar is not nothing then
+        return pVar
+    end if
+    return pDefault
+end handler
+
+handler SetRecorderProperties() returns nothing
+    if not IsAndroid() then
+        throw "Not supported on this platform"
+    end if
+
+    if mRecorder is nothing then
+        return 
+    end if
+
+    variable tRecordInput as String
+    put GetOptionalProperty(mRecordInput, "Mic") into tRecordInput
+    variable tRecordFormat as String
+    put GetOptionalProperty(mRecordFormat, "ThreeGPP") into tRecordFormat
+    variable tRecordCompressionType as String
+    put GetOptionalProperty(mRecordCompressionType, "AMR NB") into tRecordCompressionType
+    
+    _JNI_MediaRecorderSetRecordInput(mRecorder, kRecordInputs[tRecordInput])
+    _JNI_MediaRecorderSetRecordFormat(mRecorder, kRecordFormats[tRecordFormat])
+    _JNI_MediaRecorderSetRecordEncoder(mRecorder, kRecordCompressionTypes[tRecordCompressionType])
+end handler
+
+handler CreateRecorder() returns nothing
+    if mRecorder is nothing then
+        put _JNI_MediaRecorderConstructor() into mRecorder
+    end if
+
+    SetRecorderProperties()
+end handler
+
+/**
+Summary: Start recording an audio file, using the given filename
+
+Example:
+    androidRecorderStartRecording "recording.mp4"
+
+Parameters:
+pFileName: The filename to use when recording audio
+    
+Description:
+Use the <androidRecorderStartRecording> handler to start recording audio
+on Android using the input source, compression type and output format 
+selected using the <androidRecorderSetRecordInput>, 
+<androidRecorderSetRecordCompressionType>, and <androidRecorderSetRecordFormat>
+handlers. The default input source, compression and output format are 
+built-in microphone, AMR-NB and 3gpp respectively.
+
+References:
+androidRecorderSetRecordInput (handler), 
+androidRecorderSetRecordCompressionType (handler), 
+androidRecorderSetRecordFormat (handler)
+*/
+public handler androidRecorderStartRecording(in pFileName as String) returns String
+    if not IsAndroid() then
+        throw "Not supported on this platform"
+    end if
+    -- Record an mp4 file in the external storage directory with the given
+    -- name, (i.e. /storage/emulated/0/pFileName.mp4 or equivalent)
+
+    CreateRecorder()
+    
+    if mRecorder is nothing then
+        throw "unable to create recorder"
+    end if
+    
+    variable tFile as JString
+    put GetAbsolutePathToPublicDocument(pFileName) into tFile
```

Comment:
Is there a reason we restrict this to external storage? What if the incoming path is absolute already?

Reply:
Heh, good point... actually external storage path should probably be in an android utility library

Reply:
As long as the external storage path is accessible via specialFolderPath() and the recorder functions report a suitable error if the file can't be created; we can probably defer any such processing to user script.

Reply:
I don't think the external storage path is accessible via specialFolderPath...

Reply:
I think it should be:
```
	static final String s_external_prefix = "external ";
	public String getSpecialFolderPath(String p_name)
	{
		boolean t_external = false;
		if (p_name.startsWith(s_external_prefix))
		{
			t_external = true;
			p_name = p_name.substring(s_external_prefix.length());
		}

		try
		{
			if (p_name.equalsIgnoreCase("documents"))
			{
				if (t_external)
					return getContext().getExternalFilesDir(null).getAbsolutePath();
				else
					return getContext().getFilesDir().getAbsolutePath();
			}
			else if (p_name.equalsIgnoreCase("temporary") || p_name.equalsIgnoreCase("cache"))
			{
				if (t_external)
					return getContext().getExternalCacheDir().getAbsolutePath();
				else
					return getContext().getCacheDir().getAbsolutePath();
			}
		}
		catch (Exception e)
		{
			return "";
		}

		return "";
	}
```
Via 'external documents' - maybe? I think that's the same call you are using in LCB.

Reply:
Ah! It is not documented!